File: hash-map-set-table-(hashing)/2224.py
# Tc = O(n + p log p) → n for counting + p log p for sorting
# Sc = : O(p) → p = number of unique players

from collections import defaultdict

def findWinners(matches):
    from collections import defaultdict

    loss_count = defaultdict(int)

    for winner, loser in matches:
        # Ensure winner is in the map
        if winner not in loss_count:
            loss_count[winner] = 0
        loss_count[loser] += 1

    zero_losses = []
    one_loss = []

    for player in loss_count:
        if loss_count[player] == 0:
            zero_losses.append(player)
        elif loss_count[player] == 1:
            one_loss.append(player)

    return [sorted(zero_losses), sorted(one_loss)]

File: hash-map-set-table-(hashing)/2224.js
class Solution {
    findWinners(matches) {
        const lostCount = new Map();
        for (const [winner, loser] of matches) {
            if (!lostCount.has(winner)) {
                lostCount.set(winner, 0);
            }
            lostCount.set(loser, (lostCount.get(loser) || 0) + 1);
        }
        const notLost = [];
        const onceLost = [];
        for (const player of lostCount.keys()) {
            if (lostCount.get(player) === 0) {
                notLost.push(player);
            } else if (lostCount.get(player) === 1) {
                onceLost.push(player);
            }
        }
        return [notLost.sort((a, b) => a - b), onceLost.sort((a, b) => a - b)];
    }
}